yarn-project/end-to-end/src/e2e_persistence.test.ts

```diff
@@ -0,0 +1,184 @@
+import { EasyPrivateTokenContract } from '@aztec/noir-contracts/EasyPrivateToken';
+
+import { mkdtemp } from 'fs/promises';
+import { tmpdir } from 'os';
+import { join } from 'path';
+
+import { EndToEndContext, setup } from './fixtures/utils.js';
+
+describe('Aztec persistence', () => {
+  /**
+   * These tests check that the Aztec Node and PXE can be shutdown and restarted without losing data.
+   *
+   * There are four scenarios to check:
+   * 1. Node and PXE are started with an existing databases
+   * 2. PXE is started with an existing database and connects to a Node with an empty database
+   * 3. PXE is started with an empty database and connects to a Node with an existing database
+   * 4. PXE is started with an empty database and connects to a Node with an empty database
+   *
+   * All four scenarios use the same L1 state, which is deployed in the `beforeAll` hook.
+   */
+
+  // the test contract and account deploying it
+  let contractAddress: CompleteAddress;
+  let ownerPrivateKey: Fq;
+  let ownerAddress: CompleteAddress;
+
+  // a directory where data will be persisted by components
+  // passing this through to the Node or PXE will control whether they use persisted data or not
+  let dataDirectory: string;
+
+  // state that is persisted between tests
+  let deployL1ContractsValues: DeployL1Contracts;
+
+  let context: EndToEndContext;
+
+  // deploy L1 contracts, start initial node & PXE, deploy test contract & shutdown node and PXE
+  beforeAll(async () => {
+    dataDirectory = await mkdtemp(join(tmpdir(), 'aztec-node-'));
+
+    const initialContext = await setup(0, { dataDirectory }, { dataDirectory });
+    deployL1ContractsValues = initialContext.deployL1ContractsValues;
+
+    ownerPrivateKey = Fq.random();
+    const ownerWallet = await getUnsafeSchnorrAccount(initialContext.pxe, ownerPrivateKey, Fr.ZERO).waitDeploy();
+    ownerAddress = ownerWallet.getCompleteAddress();
+
+    const deployer = EasyPrivateTokenContract.deploy(ownerWallet, 1000n, ownerWallet.getAddress());
+    await deployer.simulate({});
+
+    const contract = await deployer.send().deployed();
+    contractAddress = contract.completeAddress;
+
+    await initialContext.teardown();
+  }, 100_000);
+
+  describe.each([
+    [
+      // ie we were shutdown and now starting back up. Initial sync should be ~instant
+      'when starting Node and PXE with existing databases',
+      () => setup(0, { dataDirectory, deployL1ContractsValues }, { dataDirectory }),
+      1000,
+    ],
+    [
+      // ie our PXE was restarted, data kept intact and now connects to a "new" Node. Initial synch will synch from scratch
+      'when starting a PXE with an existing database, connected to a Node with database synched from scratch',
+      () => setup(0, { deployL1ContractsValues }, { dataDirectory }),
+      10_000,
+    ],
+  ])('%s', (_, contextSetup, timeout) => {
+    let ownerWallet: AccountWallet;
+    let contract: EasyPrivateTokenContract;
```

Comment:
If we used the regular TokenContract instead, we should also be able to test that public state is properly restored, in case that's useful to check.